src/sdi-progress-dock.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2020-2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "sdi-progress-dock.h"
+#include "com.canonical.Unity.LauncherEntry.h"
+#include <glib/gi18n.h>
+#include <snapd-glib/snapd-glib.h>
+
+/**
+ * This class manages the progress bars in the dock for each snap
+ * being updated.
+ */
+
+struct _SdiProgressDock {
+  GObject parent_instance;
+
+  GApplication *application;
+  UnityComCanonicalUnityLauncherEntry *unity_manager;
+};
+
+G_DEFINE_TYPE(SdiProgressDock, sdi_progress_dock, G_TYPE_OBJECT)
+
+/**
+ * This callback should be connected to the `refresh-progress` signal from a
+ * #sdi_refresh_monitor object. It will receive the total number of tasks and
+ * how many have been done, and if the task has been completed, and with that
+ * will update the progress bars in the dock.
+ */
+void sdi_progress_dock_update_progress(SdiProgressDock *self, gchar *snap_name,
+                                       GStrv desktop_files,
+                                       gchar *task_description,
+                                       guint done_tasks, guint total_tasks,
+                                       gboolean task_done) {
+  if (desktop_files != NULL && total_tasks != 0) {
+    for (gchar **desktop_file = desktop_files; *desktop_file != NULL;
+         desktop_file++) {
+      // Update dock progress bar
+      g_autoptr(GVariantBuilder) builder =
+          g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+      g_variant_builder_add(
+          builder, "{sv}", "progress",
+          g_variant_new_double(done_tasks / ((gfloat)total_tasks)));
+      g_variant_builder_add(builder, "{sv}", "progress-visible",
+                            g_variant_new_boolean(!task_done));
+      g_variant_builder_add(builder, "{sv}", "updating",
+                            g_variant_new_boolean(!task_done));
+
+      g_autoptr(GVariant) values =
+          g_variant_ref_sink(g_variant_builder_end(builder));
+      unity_com_canonical_unity_launcher_entry_emit_update(
+          self->unity_manager, *desktop_file, values);
```

Comment:
Being a variant floating you pass to dbus, it should be consumed by it already... So I'd just do:

```suggestion
      unity_com_canonical_unity_launcher_entry_emit_update(
          self->unity_manager, *desktop_file, g_variant_builder_end(builder));
```

Reply:
Done.